plugins/micromega/persistent_cache.ml

```diff
@@ -128,54 +124,46 @@ module PHashtable (Key : HashedType) : PHashtable with type key = Key.t = struct
   (* We make the assumption that an acquired lock can always be released *)
 
   let do_under_lock kd fd f =
-    if lock kd fd then fun_protect f ~finally:(fun () -> unlock fd) else f ()
+    if lock kd fd then fun_protect f ~finally:(fun () -> unlock fd) else None
+
+  let fopen_in = open_in_bin
 
-  let fopen_in = open_in
+  let check_magic_number_and_version inch =
+    try
+      let magic = input_int32 inch in
+      let version = input_int32 inch in
+      magic = ObjFile.magic_number && version = Coq_config.vo_version
+    with End_of_file -> false
 
   let open_in (type a) f : a t =
     let flags = [O_RDONLY; O_CREAT] in
     let finch = openfile f flags 0o666 in
     let inch = in_channel_of_descr finch in
-    let exception InvalidTableFormat of a data Table.t in
+    let exception InvalidTableFormat  in
     let rec xload table =
       match read_key_elem inch with
-      | None -> table
-      | Some (hash, pos) -> xload (Table.add hash { pos; obj = None } table)
-      | exception e when CErrors.noncritical e -> raise (InvalidTableFormat table)
+      | None -> Some table
+      | Some (hash, pos) -> xload (Table.add hash { pos =  Some pos; obj = None } table)
+      | exception e when CErrors.noncritical e -> raise InvalidTableFormat
     in
+
     try
       (* Locking of the (whole) file while reading *)
-      let htbl = do_under_lock Read finch (fun () -> xload Table.empty) in
+      let htbl = do_under_lock Read finch (fun () ->
+          let table = Table.empty in
+          if check_magic_number_and_version inch
```

Comment:
Perhaps you should check the magic numbers outside of the lock. Otherwise, you might end up in a strange situation where you write into a file that does not even have the correct format.

But perhaps we can just ignore this issue, as it opens a whole new can of worms. That is, two different Coq versions are manipulating the cache at the same time, which does not bode well.

I guess it warrants a comment such as "This caching mechanism supports concurrent accesses, as long as all the Coq executables are identical. Some cache entries might get lost in the process, but none should become corrupted."